src/sp/transport/ipc/ipc.c

```diff
@@ -21,835 +21,17 @@
-		uint64_t len;
-
-		// Check to make sure we got msg type 1.
-		if (p->rx_head[0] != 1) {
-			rv = NNG_EPROTO;
-			goto error;
-		}
-
-		// We should have gotten a message header.
-		NNI_GET64(p->rx_head + 1, len);
-
-		// Make sure the message payload is not too big.  If it is
-		// the caller will shut down the pipe.
-		if ((len > p->rcv_max) && (p->rcv_max > 0)) {
-			rv = NNG_EMSGSIZE;
-			goto error;
-		}
-
-		// Note that all IO on this pipe is blocked behind this
-		// allocation.  We could possibly look at using a separate
-		// lock for the read side in the future, so that we allow
-		// transmits to proceed normally.  In practice this is
-		// unlikely to be much of an issue though.
-		if ((rv = nni_msg_alloc(&p->rx_msg, (size_t) len)) != 0) {
-			goto error;
-		}
-
-		if (len != 0) {
-			nni_iov iov;
-			// Submit the rest of the data for a read -- we want to
-			// read the entire message now.
-			iov.iov_buf = nni_msg_body(p->rx_msg);
-			iov.iov_len = (size_t) len;
-
-			nni_aio_set_iov(rx_aio, 1, &iov);
-			nng_stream_recv(p->conn, rx_aio);
-			nni_mtx_unlock(&p->mtx);
-			return;
-		}
-	}
-
-	// Otherwise we got a message read completely.  Let the user know the
-	// good news.
-
-	aio = nni_list_first(&p->recv_q);
-	nni_aio_list_remove(aio);
-	msg       = p->rx_msg;
-	p->rx_msg = NULL;
-	n         = nni_msg_len(msg);
-	nni_pipe_bump_rx(p->pipe, n);
-	ipc_pipe_recv_start(p);
-	nni_mtx_unlock(&p->mtx);
-
-	nni_aio_set_msg(aio, msg);
-	nni_aio_finish_sync(aio, 0, n);
-	return;
-
-error:
-	while ((aio = nni_list_first(&p->recv_q)) != NULL) {
-		nni_aio_list_remove(aio);
-		nni_aio_finish_error(aio, rv);
-	}
-	msg       = p->rx_msg;
-	p->rx_msg = NULL;
-	nni_pipe_bump_error(p->pipe, rv);
-	// Intentionally, we do not queue up another receive.
-	// The protocol should notice this error and close the pipe.
-	nni_mtx_unlock(&p->mtx);
-
-	nni_msg_free(msg);
-}
-
-static void
-ipc_pipe_send_cancel(nni_aio *aio, void *arg, int rv)
-{
-	ipc_pipe *p = arg;
-
-	nni_mtx_lock(&p->mtx);
-	if (!nni_aio_list_active(aio)) {
-		nni_mtx_unlock(&p->mtx);
-		return;
-	}
-	// If this is being sent, then cancel the pending transfer.
-	// The callback on the tx_aio will cause the user aio to
-	// be canceled too.
-	if (nni_list_first(&p->send_q) == aio) {
-		nni_aio_abort(&p->tx_aio, rv);
-		nni_mtx_unlock(&p->mtx);
-		return;
-	}
-	nni_aio_list_remove(aio);
-	nni_mtx_unlock(&p->mtx);
-
-	nni_aio_finish_error(aio, rv);
-}
-
-static void
-ipc_pipe_send_start(ipc_pipe *p)
-{
-	nni_aio *aio;
-	nni_msg *msg;
-	int      nio;
-	nni_iov  iov[3];
-	uint64_t len;
-
-	if (p->closed) {
-		while ((aio = nni_list_first(&p->send_q)) != NULL) {
-			nni_list_remove(&p->send_q, aio);
-			nni_aio_finish_error(aio, NNG_ECLOSED);
-		}
-		return;
-	}
-	if ((aio = nni_list_first(&p->send_q)) == NULL) {
-		return;
-	}
-
-	// This runs to send the message.
-	msg = nni_aio_get_msg(aio);
-	len = nni_msg_len(msg) + nni_msg_header_len(msg);
-
-	p->tx_head[0] = 1; // message type, 1.
```

Comment:
I think you may have overlooked this.  The wire protocol for IPC is slightly different than the other streaming protocols, because it adds another field (the message type).  If this change is not handled, it will no longer be compatible with other SP implementations for IPC.

Reply:
> I think you may have overlooked this.
> ...
> it will no longer be compatible with other SP implementations for IPC.

You are right, I got too carried away :)